### PR TITLE
Implement WebSocket extension + permessage-deflate support

### PR DESCRIPTION
**Pull request by mtrudel**
~Waiting on #38, which is in turn waiting on phoenixframework/phoenix#5030 and the work therein.~ Ready for review. @moogle19 care to have a gander?

### CODE REVIEW

**Review by moogle19 — CHANGES_REQUESTED** [review]
Nice work! Apart from 2 minor things, this looks awesome.

File: lib/bandit/websocket/permessage_deflate.ex [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    inflate_context = :zlib.open()
+    :ok = :zlib.inflateInit(inflate_context, -instance.client_max_window_bits)
+    deflate_context = :zlib.open()
+
+    :ok =
+      :zlib.deflateInit(
+        deflate_context,
+        :default,
+        :deflated,
+        -instance.server_max_window_bits,
```
[the remark this answers is not in the document]
Comment (moogle19) [reply lib/bandit/websocket/permessage_deflate.ex:96]: > Due to a known bug in the underlying zlib library, WindowBits values 8 and -8 do not work as expected. In zlib versions before 1.2.9 values 8 and -8 are automatically changed to 9 and -9. From zlib version 1.2.9 value -8 is rejected causing zlib:deflateInit/6 to fail (8 is still changed to 9). It also seem possible that future versions of zlib may fix this bug and start accepting 8 and -8 as is.

From [zlib:deflateInit/6](https://www.erlang.org/doc/man/zlib.html#deflateInit-6)

File: lib/bandit/websocket/permessage_deflate.ex [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    deflated_size = byte_size(deflated_data) - 4
+
+    deflated_data =
+      case deflated_data do
+        <<deflated_data::binary-size(deflated_size), 0x00, 0x00, 0xFF, 0xFF>> -> deflated_data
+        deflated -> deflated
+      end
+
+    if context.server_no_context_takeover, do: :zlib.deflateReset(context.deflate_context)
+    {:ok, deflated_data, context}
```
[the remark this answers is not in the document]
Comment (mtrudel) [reply lib/bandit/websocket/permessage_deflate.ex:133]: ibid.

File: lib/bandit/websocket/permessage_deflate.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (mtrudel) [reply lib/bandit/websocket/permessage_deflate.ex:96]: Most excellent catch! I think we do need this. I'll add it.

**Comment by mtrudel** [comment]
thanks for the review @moogle19 !

**Pull request merged**